feat: Selective CSP header stripping from HTTPResponse

pgoforth opened the pull request:
- Closes [#1030](https://github.com/cypress-io/cypress/issues/1030)

### User facing changelog

- Cypress now allows selective stripping of `Content-Security-Policy` and `Content-Security-Policy-Report-Only` header directives via the `stripCspDirectives` config option.

### Additional details
- add generated `nonce` to inline script injection
- append generated `nonce` policy value to each CSP header `script-src-elem`, `script-src`, and `default-src` directive if provided in original response
- remove mandatory `content-security-policy` and `content-security-policy-report-only` header stripping
- selectively strip problematic CSP directive `frame-ancestors` because it prevents Cypress from loading target into iframe
- allows for proviiding CSP directives to strip in the `parseCspHeaders` method
- add config option `stripCspDirectives` that permits selective stripping of individual CSP directives
- default value for the `stripCspDirectives` config option maintains existing CSP header stripping

### Steps to test

1. Create a local HTTP server that responds to a request with CSP headers defined
2. Run a Cypress test that sends a request to the local server, and tests the response header, similar to the following:
```javascript
const route = 'http://localhost:8080';
cy.request(route).as('route');
cy.get('@route').should(response => {
    expect(response).to.have.property('headers');
    expect(response.headers).to.have.property('content-security-policy');
    expect(response.headers['content-security-policy']).to.match(
        /script-src 'nonce-[^-A-Za-z0-9+/=]|=[^=]|={3,}'/
    );
});
cy.visit(route);
```

### How has the user experience changed?
This change does not affect UI/UX

### PR Tasks
- [X] Have tests been added/updated?
- [X] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?
[Documentation for `stripCspDirectives`](https://github.com/cypress-io/cypress-documentation/pull/5194)
- [X] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

pgoforth committed:
feat: Selective CSP header directive stripping from HTTPResponse
- uses `stripCspDirectives` config option

pgoforth commented:
I'd like to use this time while the PR is in draft to discuss _potentially_ exposing a configuration option allowing the selective stripping of certain CSP directives, with the goal being to `default` strip to ALL directives (which maintains parity with current version of Cypress)

The only reason to consider this is to make this feature opt-in, with an eye to changing the default in a future release. I would most definitely need some guidance as to how to best implement any global config setting, but in the end, we may not even want to expose a config property at all. I'll leave it up for discussion.

**EDIT**
I decided it's in the best interest of all Cypress users to create a config flag called `stripCspDirectives: boolean | string | string[]` and default it to `true`

`stripCspDirectives: true` Would remove the CSP headers entirely and maintain parity with old versions of Cypress.

`stripCspDirectives: false` would strip ONLY functionality breaking directives...like `frame-ancestors`

`stripCspDirectives: '<directive-name>'` would strip ONLY the specific directive defined, potentially including functionality breaking directives...like `frame-ancestors`

`stripCspDirectives: ['<directive-name>', '<directive-name>', ...]` would strip ONLY the specific directives defined in the array, potentially including functionality breaking directives...like `frame-ancestors`

pgoforth commented:
Implementation for the config parameter is blocked by https://github.com/cypress-io/cypress/issues/21151. It seems that the server restart is either:

1. Not happening
2. Not getting the config updates

I would imagine that it's the later.

**EDIT**
Seems the restart never happens, which would certainly be the reason for https://github.com/cypress-io/cypress/issues/21151 because the config in the server context is static

pgoforth commented:
Everything here works when you set the `stripCspDirectives` attribute in you `cypres.config.js` file...you just cannot set is on a per spec/run basis using `Cy.config`. This is the same fate as the `blockHosts` parameter in https://github.com/cypress-io/cypress/issues/21151.

This PR stands on it's own, and hopefully (since the parameter is setup the same way as `blockHosts`, when that issue is resolved, this one should be too.

mschile commented:
Hi @pgoforth 👋, thanks for this contribution! I'm planning on bringing it up next week with the team so we can discuss it.

chrisbreiding commented on cli/types/cypress.d.ts:
```diff
@@ -3048,6 +3048,14 @@ declare namespace Cypress {
      * @default 'top'
      */
     scrollBehavior: scrollBehaviorOptions
+    /**
+     * Indicates whether Cypress should strip CSP header directives from the application under test.
+     * NOTE: When this option is set to `false`, Cypress **will** continue to strip the `frame-ancestors` directive
+     * because it prevents loading the target document into an iframe. Please see the documentation for more information.
+     * @see https://on.cypress.io/configuration#stripCspDirectives
+     * @default true
+     */
+    stripCspDirectives: boolean | string | string[],
```
```suggestion
    /**
     * Indicates whether Cypress should strip CSP header directives from the application under test.
     * NOTE: Setting this value to 'none' or not stripping `frame-ancestors`
     * may prevent Cypress from functioning because its presence prevents 
     * loading the target document into an iframe. Please see the documentation 
     * for more information.
     * @see https://on.cypress.io/configuration#stripCspDirectives
     * @default 'all'
     */
    stripCspDirectives: 'all' | 'none' | string | string[],
```

A couple changes: 

- Use only strings with some enumerated values instead of mixing booleans and strings. `true` is now "all", `false` is now "none"
- It seems inconsistent to strip `frame-ancestors` when false/none is specified, but not when specific directives are specified. I'd rather give users control of this and have the caveat be consistent amongst non-"all" values.

pgoforth replied:
@chrisbreiding So what you are saying is that 'none' should allow all CSP headers...even the ones that prevent Cypress from running? If that's the case, then why 'none' and not an empty string or empty array?

As an adendum to that question, why allow a static attribute like 'none' to implement behavior that blocks Cypress from running on sites with 'frame-ancestors'? Is it because it would error people's tests immediately and force them to specify 'frame-ancestors'?

Edit:
Maybe I can explain it better this way. I agree with the idea of eliminating booleans, that's a great idea. I do believe, however, there should be an enum option that _only_ strips problematic CSP headers and does not require the end user to explicitly identify `frame-ancestors`.

Imagine that there is a new CSP directive that is released that inhibits Cypress functionality. That would require all users to now add a new directive to their `stripCspDirectives` array. If there were an enum that automatically stripped the problematic directives, Cypress could add the directive to the `unsupportedCSPDirectives`, and all users who used the enum value would get that header stripped.

I think the enum value "none" is probably not a good one for this behavior...and naming things *IS* hard. How about some of these options?

- "essential"
- "critical"
- "unsupported"
- "blocking"
- "breaking"
- "functional"
- "obstructive"
- "disruptive"

Now that I've written them all, I think I like **"breaking"** the best.

chrisbreiding replied:
> Imagine that there is a new CSP directive that is released that inhibits Cypress functionality. That would require all users to now add a new directive to their stripCspDirectives array. If there were an enum that automatically stripped the problematic directives, Cypress could add the directive to the unsupportedCSPDirectives, and all users who used the enum value would get that header stripped.

I hadn't thought of that. I agree we should do away with "none" and have an enum value that equates to what `false` does. What about "minimum" since it would be stripping the minimum number of directives for Cypress to function?

pgoforth replied:
@chrisbreiding Sounds good. I'll update using "minimum"

chrisbreiding commented on cli/CHANGELOG.md:
```diff
@@ -47,6 +47,8 @@ _Released 04/17/2023_
 
 **Features:**
 
+- Cypress now allows targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive stripping from requests via the [`stripCspDirectives`](https://docs.cypress.io/guides/references/configuration#stripCspDirectives) config option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030). Addressed in [#26483](https://github.com/cypress-io/cypress/pull/26483).
```
Since we've had a couple releases in the interim, can you move this up under a new `12.3.0` heading, release date `05/23/2023`? I promise we'll do everything we can to get this merged in time for that release 😅

chrisbreiding commented on packages/config/src/validation.ts:
```diff
@@ -41,6 +41,33 @@ const isFalse = (value: any): boolean => {
+  }
+}
+
+export const validateAll = (...validations: ValidationFn[]): ValidationFn => {
+  return (key: string, value: any): ValidationResult => {
+    return validations.reduce((result: ValidationResult, validation: ValidationFn) => {
+      if (result !== true) {
+        return result
+      }
+
+      return validation(key, value)
+    }, true)
+  }
+}
+
```
```suggestion
```

I don't see this used anywhere besides tests, so I'd rather not add it.

pgoforth replied:
Understood. Additionally, by forcing the flag to be a string or string[], you eliminate the need for this helper AND `validateAny`, so both can be removed.

Edit:
Actually, the migration to an `ENUM` + `string[]` still necessitates `validateAny`. I will remove `validateAll`.

chrisbreiding commented:
This looks good, except that the `run-app-component-tests-chrome` tests are failing. Doesn't seem like the changes here should be impacting those tests, but they are failing consistently on this PR, so it doesn't look like flake.

pgoforth commented:
@chrisbreiding I will investigate the failing tests

Edit:
@chrisbreiding I performed a rebase on the PR branch and ran `yarn workspace @packages/app cypress:run:ct --browser chrome` locally and got everything passing. Is there a specific command I sould be running instead? Maybe the rebase fixed it.
![Screenshot 2023-05-12 at 11 46 36 AM copy](https://github.com/cypress-io/cypress/assets/1943491/d4f718e5-7817-40a5-ab3f-14a03bfeabdc)

chrisbreiding reviewed (approved):
Looks good! Thanks for the contribution

AtofStryker commented on cli/types/cypress.d.ts:
```diff
@@ -3048,6 +3048,18 @@ declare namespace Cypress {
+     *
+     * Please see the documentation for more information.
+     * @see https://on.cypress.io/configuration#stripCspDirectives
+     * @default true
```
```suggestion
     * @default 'all'
```

AtofStryker commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -302,6 +304,24 @@ const SetInjectionLevel: ResponseMiddleware = function () {
   }
 
+  // If the user has specified CSP directives to strip, we must remove them from the CSP header
+  const stripDirectives = this.config.stripCspDirectives === 'minimum' ? unsupportedCSPDirectives : this.config.stripCspDirectives
```
Can we move this to the `OmitProblematicHeaders` middleware? Even though it occurs after this middleware, I think it should be fine since we are only performing the calculation if injecting here and the stripCSP option isn't `all`. Only reason is to make it clearer that we are still stripping these headers out if configured to do so

pgoforth replied:
@AtofStryker Just to be clear, we are talking about moving the header stripping to the `OmitProblematicHeaders`, and leaving nonce injection where it currently is?

pgoforth replied:
@AtofStryker I can move the directive stripping to `OmitProblematicHeaders`, but we will be replicating a lot of code.

Stripping all the specified CSP directives happens in three steps:
1. Parse the CSP headers into policy maps of directive:value pairs
2. Remove the stripped directives from each header policy map
3. Apply the modified policy map back to the CSP header it was parsed from

When we add the nonce value in `SetInjectionLevel` it is only added when we need it, and it has to be added to each header that has a matching directive. This mean we have to repeat steps 1 and 3 but modify step 2
1. Parse the CSP headers into policy maps of directive:value pairs
2. Append the nonce to necessary directives in the policy map
3. Apply the modified policy map back to the CSP header it was parsed from

The reason everything is currently in `SetInjectionLevel` is because we would not have to repeat steps 1 and 3

Edit:
I'm reconfiguring the code now. There's a little replication, but it comes from common helper methods, so I'm not concerned. I'm going to have to move a bunch of tests around, but I think that moving the stripping into the other middleware makes more sense conceptually.

AtofStryker replied:
@pgoforth I'm only referring to the header stripping in `OmitProblematicHeaders` . Keeping the directive stripping in `SetInjectionLevel` makes sense to me, unless its harder to tell if the header was previously stripped, which we can just leave the way it is and just commit in the `OmitProblematicHeaders` code that CSP is handled in SetInjection and may be stripped out depending on `stripCspDirectives`

AtofStryker commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -311,6 +331,51 @@ const SetInjectionLevel: ResponseMiddleware = function () {
+
+      // Iterate through each CSP header
+      incomingCspHeaders.forEach((headerName) => {
+        const excludeDirectives = _.castArray(stripDirectives)
```
do we need to `_.castArray` here since `stripDirectives` should be an array at this point since we know it isn't `'all'`?

pgoforth replied:
Not anymore. It used to be of type `boolean|string|string[]` and you could add directives as individual strings. This can   be removed.

AtofStryker commented on packages/proxy/lib/http/util/csp-header.ts:
```diff
@@ -0,0 +1,74 @@
+   * and prevents the iframe content from being loaded if it detects that it is not being loaded in the
+   * top-level frame.
+   */
+  'frame-ancestors',
```
I wonder if in the future we can support `self` here for documents nested inside the Application Under Test Iframe since we should only be really about `frame-ancestors 'none'` (assuming I understand that correctly?)

pgoforth replied:
I think that yes, in the future this could be modified to not strip entire directives, but directive:value combinations only. The reason I would be hesitant is that CSP is inclusive by nature. Having a policy that says:
```
Content-Security-Policy: frame-ancestors 'self' 'https://fakedomain.com'
```

...would mean we allow iframes from same origin, and `https://fakedomain.com`. So excluding `self` would mean that we **only** allow iframe from the `https://fakedomain.com` origin.

What we MIGHT be able to do is, instead of removing the directive, augment the `frame-ancestors` directive and supply it with the origin of the Cypress run...the same way we are supplying the `nonce` to the script source directives.

AtofStryker commented on packages/proxy/lib/http/util/inject.ts:
```diff
@@ -21,13 +23,17 @@ export function partial (domain, options: InjectionOpts) {
   return oneLine`
-    <script type='text/javascript'>
+    <script type='text/javascript'${
+      cspNonce ? ` nonce="${cspNonce}"` : ''
```
since we use this three places, can we refactor this to a function that returns the appropriate string when the `cspNonce` is passed in?

AtofStryker reviewed (changes requested):
Sorry for the delay on a review and feedback @pgoforth. I only had a few minor things but otherwise looks great! Thank you for adding thorough tests and making the functionality passive by default, as well as your patience with #1030. Big contribution 🎉

pgoforth commented on packages/driver/cypress/e2e/e2e/csp-headers.cy.js:
```diff
@@ -0,0 +1,62 @@
+describe('csp-headers', () => {
+  // Currently unable to test spec based config values for stripCspDirectives
+  if (cy.config('stripCspDirectives') === 'all') {
```
Since we cannot set the `stripCspDirectives` on a per spec/test basis due to https://github.com/cypress-io/cypress/issues/21151, I've added both tests here for stripping the CSP header, and not stripping, and running them through an `if` statement. In the future, we can run both these tests and set the config value for each test.

pgoforth commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -345,6 +348,41 @@ const SetInjectionLevel: ResponseMiddleware = function () {
+
+    // In order to allow the injected script to run on sites with a CSP header
+    // we must add a generated `nonce` into the response headers
+    const nonce = crypto.randomBytes(16).toString('base64')
+
+    // Iterate through each CSP header
+    cspHeaderNames.forEach((headerName) => {
+      const policyArray = parseCspHeaders(this.res.getHeaders(), headerName)
```
`nonce` injection is still happening in `SetInjectionLevel`, but we are going to base it off `this.res.getHeaders()` instead of `this.incomingRes.headers` so that we can pick up the stripped directives in `OmitProblematicHeaders`.

pgoforth commented:
@AtofStryker I think I've hit all the points that need addressing.

I have moved all CSP header/directive stripping into `OmitProblematicHeaders`, and left the `nonce` injection in `SetInjectionLevel`. Also added a few more tests and a couple docs on the more convoluted new methods.

AtofStryker commented:
@pgoforth thank you for going through an addressing my feedback and sorry for the delay! I've been talking through this feature with the team and there are a few things we want to change:
* instead of having a `stripCspDirectives` config option, want to make this an experimental flag called `experimentalCspAllowList` which relaxes the CSP headers that are stripped. When enabled, it does not strip the CSP directives we know that likely won't cause issues with Cypress. This also gives us some flexibility when making changes to the feature before we make it generally available. Here is that list:
  * `child-src`, `connect-src`, `font-src`, `frame-src`, `img-src`, `manifest-src`, `media-src`, `object-src`, `prefetch-src`,` script-src-attr`, `worker-src`, `base-ure`, `report-uri`, `report-to`, `upgrade-insecure-requests`
* `experimentalCspAllowList=false` - the default and strips everything like Cypress does today.
* `experimentalCspAllowList=true` - does not strip CSP directives listed above. This will only strip directives that can cause problems or we know for sure will cause problems with Cypress 
    * directives that might cause a problem: `default-src`, `script-src`, and `script-src-elem`, `sandbox`, `form-action`, `navigate-to`
       * these are the possible "allowed" directives users can opt-in to
    * directives that definitely will cause a problem: `frame-ancestors`, `trusted-types`, `require-trusted-types-for`.
      * We always strip this out and users cannot skip stripping these (similar to your 'minimum' config option with trusted types added in).

* When the user defines an allow list like, `experimentalCspAllowList=['default-src', 'script-src' , 'script-src-elem']`,  the a user opting out of stripping these additional directives. With this option set, the nonce injection code you wrote will runs to make sure the inline driver script runs.
  * The main reason we want this to be opt in is there may be cases where these directives don't work, in particular `SHA` hashing over Nonce in directives where Cypress rewrites the inline script to prevent framebusting and the `SHA` breaks so the script doesn't run
![Screenshot 2023-05-19 at 11 20 51 AM](https://github.com/cypress-io/cypress/assets/3980464/dc397876-0259-4c43-a096-1605d594960b)
* if the use specifies `frame-ancestors`, `trusted-types`, or `require-trusted-types-for` as an allowed , the configuration validation will fail and throw an error. 

The nice part with the experimental is it gives us some flexibility to possibly iterate on this in the future without worrying significantly about breaking the API

It shouldn't take too much effort to get the PR and docs updated to reflect these changes, hopefully 😄 . Are you willing to update your PR with these changes?

emilyrohrbough commented:
@pgoforth Can you also share what you are trying to accomplish with this change? It wasn't clear your desired end-goal since it is possible to verify these CSP headers exist on the requests.

pgoforth commented:
@AtofStryker 
I will be able to address these items next week. What I will likely do is maintain a lot of the underlying code that handles the directive stripping and simply rename config parameters/values, and reverse the matching.

@emilyrohrbough 
The goal of this feature is to be able to test that websites are delivering and handling CSP appropriately. I should be able to run an E2E test against a website and validate that:
1. My backend delivered the CSP headers appropriately

2. My frontend handles the headers appropriately ie.
 - I'm generating nonces correctly
 - My CSP policies are preventing the things I expect them to prevent

3. As a third party developer, I can test that I have appropriate fallbacks in place to handle these scenarios. For example, if my code is deployed on a site where the CSP headers are outside of my control.

Those are just a few examples. Currently I cannot automate those types of tests through Cypress.

Edit:
@AtofStryker out of curiouslity (after looking at the `experimentalModifyObstructiveThirdPartyCode` flag) have you all tried using an iframe `sandbox` attribute to prevent access of `window.top`, and/or an `onbeforeunload` handler on the iframe container? If so, what were the obsticles involved? You can direct me to a PR conversation if it was already discussed there.

Edit2:
As an asside, could we not also create a TrustedPolicy using the `nonce` injected script tag and add that policy into the `trusted-types` directive? The policy would simply verify that the injections are coming from Cypress, and wrap them as trusted. That would eliminate the need to forcefully strip `trusted-types` and `require-trusted-types-for`. That seems like it would be a totally new feature on a different PR, but I believe the work in this PR would pave the way for that. Thoughts?

AtofStryker commented:
> @AtofStryker out of curiouslity (after looking at the experimentalModifyObstructiveThirdPartyCode flag) have you all tried using an iframe sandbox attribute to prevent access of window.top, and/or an onbeforeunload handler on the iframe container? If so, what were the obsticles involved? You can direct me to a PR conversation if it was already discussed there.

I don't believe we have tried that, but I don't think it would work because most the code that framebusts usually will short circuit and prompt a top redirect, where as the `sandbox` attr only solves the later problem. We still need the rest of that code to run for the site to behave normally. 

> As an asside, could we not also create a TrustedPolicy using the nonce injected script tag and add that policy into the trusted-types directive? The policy would simply verify that the injections are coming from Cypress, and wrap them as trusted. That would eliminate the need to forcefully strip trusted-types and require-trusted-types-for. That seems like it would be a totally new feature on a different PR, but I believe the work in this PR would pave the way for that. Thoughts?

That is something we can do to at least fix some of the injection sink issues that would be present if trusted-types CSP is configured. There are likely other cases we are missing, so a full audit of the app for things `trusted-types` CSP would break would need to be performed focusing on [these bullet points](https://web.dev/trusted-types/#api-introduction), then an implementation to address those areas that only uses the trusted functions we inject to perform those operations if configured. It could be a lot of effort to support something that is experimental tech, but it would be possible as a different feature and this PR would help pave a way to that if one is willing to take a stab at it.

pgoforth committed:
feat: Selective CSP header directive permission from HTTPResponse
- uses `experimentalCspAllowList` config option

pgoforth commented:
@AtofStryker 
Updates have been pushed, tests rewritten, and I've tested the build against our own usage of CSP headers and the tests I wrote against it all pass and fail appropriately. This is ready for review again.

Edit:
Had to move the `CHANGELOG` to the next release to get automation running

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -1,4 +1,12 @@
+
+**Features:**
+
+- Cypress now has a targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#experimentalCspAllowList) config option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030). Addressed in [#26483](https://github.com/cypress-io/cypress/pull/26483).
```
```suggestion
- Cypress now has a targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#experimentalCspAllowList) configuration option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030).
```

pgoforth replied:
@AtofStryker I removed the "Addressed" item, but the Semantic PR automation yelled at me. Just letting you know.
I also updated the docs link to properly deep link.

AtofStryker commented on cli/types/cypress.d.ts:
```diff
@@ -3048,6 +3050,19 @@ declare namespace Cypress {
+     * specified will remain in the response headers.
+     *
+     * Please see the documentation for more information.
+     * @see https://on.cypress.io/configuration#experimentalCspAllowList
```
note to myself that we will need to deploy the on links (see release process [step 2](https://github.com/cypress-io/cypress/blob/develop/guides/release-process.md#before-publishing-a-new-version))for this to work. I will take care of creating the manifest PR for this next week

AtofStryker replied:
@mschile since this isn't making it into today's release you may have to handle it for the next release

AtofStryker commented on cli/types/cypress.d.ts:
```diff
@@ -3247,14 +3262,14 @@ declare namespace Cypress {
 
   interface SuiteConfigOverrides extends Partial<
-    Pick<ConfigOptions, 'animationDistanceThreshold' | 'blockHosts' | 'defaultCommandTimeout' | 'env' | 'execTimeout' | 'includeShadowDom' | 'numTestsKeptInMemory' | 'pageLoadTimeout' | 'redirectionLimit' | 'requestTimeout' | 'responseTimeout' | 'retries' | 'screenshotOnRunFailure' | 'slowTestThreshold' | 'scrollBehavior' | 'taskTimeout' | 'viewportHeight' | 'viewportWidth' | 'waitForAnimations'>
+    Pick<ConfigOptions, 'animationDistanceThreshold' | 'blockHosts' | 'defaultCommandTimeout' | 'env' | 'execTimeout' | 'experimentalCspAllowList' | 'includeShadowDom' | 'numTestsKeptInMemory' | 'pageLoadTimeout' | 'redirectionLimit' | 'requestTimeout' | 'responseTimeout' | 'retries' | 'screenshotOnRunFailure' | 'slowTestThreshold' | 'scrollBehavior' | 'taskTimeout' | 'viewportHeight' | 'viewport ...
```
we shouldn't need to include the type here in the `SuiteConfigOverrides` or the `TestConfigOverrides` since you have it added to the `ResolvedConfigOptions`

AtofStryker commented on packages/config/__snapshots__/validation.spec.ts.js:
```diff
@@ -19,208 +19,233 @@ Expected at least one browser
-  "key": "mockConfigKey",
-  "value": "",
-  "type": "a fully qualified URL (starting with `http://` or `https://`)"
+  'key': 'mockConfigKey',
```
I think the double quote to single quote is a side effect of the husky hook 🥴

AtofStryker commented on packages/config/src/validation.ts:
```diff
@@ -148,6 +163,33 @@ export const isOneOf = (...values: any[]): ((key: string, value: any) => ErrResu
+  const validValues = values.map((a) => str(a)).join(', ')
+
+  return (key, value) => {
+    if (!Array.isArray(value)) {
```
Does it make sense for these to return the same error message if we were expecting an array of configuration? It might help guide the user a bit more accurately. Otherwise, we can just condense the logic into one statement and remove the other `if` block.
```suggestion
    if (!Array.isArray(value) || !value.every((v) => values.includes(v))) {
```

AtofStryker commented on packages/config/test/validation.spec.ts:
```diff
@@ -388,4 +421,52 @@ describe('config/src/validation', () => {
+      let msg = validateFail(key, value)
+
+      expect(msg).to.not.be.true
+      snapshot('not a an array error message', msg)
```
```suggestion
      snapshot('not an array error message', msg)
```

AtofStryker commented on packages/driver/cypress/e2e/e2e/csp-headers.cy.js:
```diff
@@ -0,0 +1,62 @@
+describe('csp-headers', () => {
+  // Currently unable to test spec based config values for experimentalCspAllowList
+  if (cy.config('experimentalCspAllowList') === false || cy.config('experimentalCspAllowList') === true) {
```
This test looks conditional based on whatever `experimentalCspAllowList` is configured to. To get more coverage, we might want to move these to a `system-test`. That requires some domain knowledge so I can work on getting these moved over

pgoforth replied:
The original test was written with the idea that you could change the `experimentalCspAllowList` on a per spec basis. When I encountered https://github.com/cypress-io/cypress/issues/21151 I realized that the proxy would never get updated config values, so it became impossible to alter this value after initialization. I left the tests, but wrapped them in if statements as an example of how to handle both cases. If you know a way to integrate both through another means, that's preferred.

AtofStryker replied:
the system test should allow us to integrate each case in this test to afford us more test coverage in testing this cases. When I get a chance I will push a commit to your branch refactoring this out into system tests

AtofStryker replied:
I pushed e4780f82be to your most recent rebased branch which should refactor this out into system tests and more or less test the same behavior

AtofStryker commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -735,8 +800,8 @@ export default {
   AttachPlainTextStreamFn,
   InterceptResponse,
   PatchExpressSetHeader,
+  OmitProblematicHeaders, // Since we might modify CSP headers, this middleware needs to come BEFORE SetInjectionLevel
```
I'd like to move the order the middleware is defined in this file to reflect the export/execution order, but it might make the PR harder to review and think I can do it in a follow up.

pgoforth replied:
Agreed, that's a lot of churn.

AtofStryker commented on packages/proxy/lib/http/response-middleware.ts:
```diff
@@ -403,13 +441,39 @@ const OmitProblematicHeaders: ResponseMiddleware = function () {
 
   this.res.set(headers)
 
+  if (!this.config.experimentalCspAllowList) {
```
can we inverse this logic, like:

```
if (this.config.experimentalCspAllowList) {
    // allow CSP directives and check configured allow list
} else {
    /// strip the CSP headers
}

AtofStryker reviewed (changes requested):
Looking good! Just a few things.

It's a bit hidden, but would you be able to add an i18n description for the feature in [en-US](https://github.com/cypress-io/cypress/blob/develop/packages/frontend-shared/src/locales/en-US.json#L507)? It is used when we show the experimental flags in the app in `open` mode.

The driver test added seems to flex on the CSP configuration. If we want to test all outcomes, we can add a `system-test` which seems like the best direction. I will work on adding this 
 as it is somewhat difficult for an open source contributor to get set up and running correctly.(though I am on vacation starting EOD tomorrow-Tuesday so it might take a few days)

pgoforth committed:
Address Review Comments:
- Add i18n for `experimentalCspAllowList`
- Remove PR link in changelog
- Fix docs link in changelog
- Remove extra typedef additions
- Update validation error message and snapshot
- Fix middleware negated conditional

AtofStryker committed:
chore: refactor driver test into system tests to get better test
coverage on experimentalCspAllowList options

AtofStryker commented on packages/data-context/src/sources/migration/legacyOptions.ts:
```diff
@@ -86,6 +86,10 @@ const resolvedOptions: Array<ResolvedConfigOption> = [
+  }, {
+    name: 'experimentalCspAllowList',
+    defaultValue: false,
+    canUpdateDuringTestTime: true,
```
I am thinking we may not want users to update this config during test time as it could lead to inconsistent results between tests.
```suggestion
    canUpdateDuringTestTime: false,
```

pgoforth replied:
Since this doesn't actually work due to https://github.com/cypress-io/cypress/issues/21151 we can change this no problem.

AtofStryker reviewed (commented):
Got the system tests added in e4780f82be. I am going to re review this tomorrow but I am thinking we are in a good spot!

AtofStryker commented on cli/CHANGELOG.md:
```diff
@@ -1,4 +1,12 @@
+
+**Features:**
+
+- Cypress now has a targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#Experimental-Csp-Allow-List) configuration option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030).
```
I think we might need the 'Addressed in' for features to make the changelog job happy (sorry if I said otherwise 🥴 )
```suggestion
- Cypress now has a targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#Experimental-Csp-Allow-List) configuration option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030). Addressed in [#26483](https://github.com/cypress-io/cypress/pull/26483)
```

AtofStryker reviewed (commented):
I'll take a look why the build job is failing tomorrow

pgoforth committed:
Address Review Comments:
- Remove legacyOption for `experimentalCspAllowList`
- Update App desc for `experimentalCspAllowList` to include "Content-Security-Policy-Report-Only"
- Modify CHANGELOG wording
- Specify “never” overrideLevel
- Remove unused validator (+2 squashed commits)
- Add "Addresses" note in CHANGELOG to satisfy automation
- Set `canUpdateDuringTestTime` to `false` to prevent confusion

pgoforth commented:
@AtofStryker I updated those two remaining issues. Thanks for migrating those tests.

AtofStryker commented:
@pgoforth the only hold up for review is currently the CI builds not running on contributor PRs, which I am working on addressing in #26897. Once its working and fixed, I will merge and and we can get these tests to run in CI and see what issues there may be.

emilyrohrbough commented on packages/data-context/src/sources/migration/legacyOptions.ts:
```diff
@@ -86,6 +86,10 @@ const resolvedOptions: Array<ResolvedConfigOption> = [
     name: 'exit',
     defaultValue: true,
     canUpdateDuringTestTime: false,
+  }, {
+    name: 'experimentalCspAllowList',
+    defaultValue: false,
+    canUpdateDuringTestTime: false,
```
this isn't a legacy config option so we dont want to add this here.

pgoforth replied:
Will remove.

emilyrohrbough commented on packages/frontend-shared/src/locales/en-US.json:
```diff
@@ -504,6 +504,10 @@
       "description": "If you'd like to try out new features that we're working on, you can enable beta features for your project by turning on the experimental features you'd like to try. {0}",
+      "experimentalCspAllowList": {
+        "name": "CSP Allow List",
+        "description": "Enables Cypress to selectively permit Content-Security-Policy header directives, including those that might otherwise block Cypress from running."
```
do we need to mention `Content-Security-Policy-Report-Only` here?

pgoforth replied:
Good call. I'll add it.

AtofStryker commented on packages/config/src/options.ts:
```diff
@@ -198,6 +198,12 @@ const driverConfigOptions: Array<DriverConfigOption> = [
+  }, {
+    name: 'experimentalCspAllowList',
+    defaultValue: false,
+    validation: validate.validateAny(validate.isBoolean, validate.isSubsetOf('script-src-elem', 'script-src', 'default-src', 'sandbox', 'form-action', 'navigate-to')),
```
Is `validateAny` doing the same thing here as `isOneOf`? We might not need a new function which would be nice 😄 .

```suggestion
    validation: validate.isOneOf(validate.isBoolean, validate.isSubsetOf('script-src-elem', 'script-src', 'default-src', 'sandbox', 'form-action', 'navigate-to')),
```

pgoforth replied:
@AtofStryker `isOneOf` is not the same as `validateAny`
`validateAny` takes additional validators as arguments, and runs the validators against the value, and if ANY of the validators pass, then the validation passes.

AtofStryker replied:
ah that makes sense. Looking closer at it is looks like you cant really pass a validator in that has already been executed.

pgoforth replied:
That is correct. It accepts validator functions that are then run in sequence against the value provided to the configuration option. In this case `isSubsetOf` is a thunk that returns a validator.

AtofStryker commented on packages/config/src/validation.ts:
```diff
@@ -321,16 +359,24 @@ export function isFullyQualifiedUrl (key: string, value: any): ErrResult | true
   )
 }
 
+export function isArrayOfStrings (key: string, value: any): ErrResult | true {
+  if (isStringArray(value)) {
+    return true
+  }
+
+  return errMsg(key, value, 'an array of strings')
+}
```
Do we still need this? I don't think the validator is used.
```suggestion
```

pgoforth replied:
We do not need this anymore. Removing.

emilyrohrbough commented on cli/CHANGELOG.md:
```diff
@@ -1,4 +1,12 @@
+
+**Features:**
+
+- Cypress now has a targeted Content-Security-Policy and Content-Security-Policy-Report-Only header directive allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#Experimental-Csp-Allow-List) configuration option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030). Addressed in [#26483](https://github.com/cypress-io/cypress/pull ...
```
```suggestion
- Cypress can now test pages with targeted `Content-Security-Policy` and `Content-Security-Policy-Report-Only` header directives by specifying the allow list via the [`experimentalCspAllowList`](https://docs.cypress.io/guides/references/configuration#Experimental-Csp-Allow-List) configuration option. Addresses [#1030](https://github.com/cypress-io/cypress/issues/1030).
```

emilyrohrbough commented on packages/config/src/options.ts:
```diff
@@ -198,6 +198,12 @@ const driverConfigOptions: Array<DriverConfigOption> = [
+    name: 'experimentalCspAllowList',
+    defaultValue: false,
+    validation: validate.validateAny(validate.isBoolean, validate.isSubsetOf('script-src-elem', 'script-src', 'default-src', 'sandbox', 'form-action', 'navigate-to')),
+    overrideLevel: 'any',
```
We dont want to support override this at test time,
```suggestion
    overrideLevel: 'never',
```

pgoforth replied:
You are correct...although I'd like to at some point. 😉

emilyrohrbough commented on packages/config/src/options.ts:
```diff
@@ -198,6 +198,12 @@ const driverConfigOptions: Array<DriverConfigOption> = [
+  }, {
+    name: 'experimentalCspAllowList',
+    defaultValue: false,
+    validation: validate.validateAny(validate.isBoolean, validate.isSubsetOf('script-src-elem', 'script-src', 'default-src', 'sandbox', 'form-action', 'navigate-to')),
```
isSubsetOf to me indicates you could specify all of these options. Can we rename to reflect the intent?

pgoforth replied:
@emilyrohrbough You can specify all those options.

emilyrohrbough replied:
sorry I mis-typed, subset to me indication would would not be able to

pgoforth replied:
@emilyrohrbough What would you suggest as a name?
`includesOnly`
`matchEvery`
`membersOf`
`arrayContainsOnly`
`isArrayContainingAny`

I'm bad a naming things...and accurate descriptions seem to be very long names 🤷

pgoforth replied:
Going to go with `isArrayIncludingAny`

pgoforth commented:
@AtofStryker A thought just occurred to me. Given that the `frame-src` could ALSO be supplied with a SHA hash value, would Cypress perform any rewrites on the HTML content loaded into an iframe that would invalidate the SHA?

If so, perhaps we add `frame-src` to the *optional* allowable directive list, and not just always permit it.

AtofStryker commented:
> @AtofStryker A thought just occurred to me. Given that the `frame-src` could ALSO be supplied with a SHA hash value, would Cypress perform any rewrites on the HTML content loaded into an iframe that would invalidate the SHA?
> 
> If so, perhaps we add `frame-src` to the _optional_ allowable directive list, and not just always permit it.

@pgoforth is that is the case, we should probably make it optional since Cypress may rewrite the javascript in the actual html document to prevent framebusting. Do you have any examples of a `frame-src` SHA hash? I'm guessing the hash would be the content of a given iframe?

pgoforth commented:
@AtofStryker I do not have a working example ATM, however [the MDN page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/frame-src#sources) says `frame-src` can use it, and [the W3 spec](https://www.w3.org/TR/CSP3/#directive-frame-src) also indicates that it's valid. I would expect that if the raw HTML value for the src of the iframe did not validate against that hash, it would fail CSP. I cannot verify that at present, but it does _feel_ this that would be the validation applied.

All that said, I do not see any harm in making `frame-src` optionally allowed, it just means one extra item that has to be added to your allow list if you want it.

Edit:
Also, `frame-src` has a fallback to `child-src`...so...perhaps that too.

This is part of the reason why the first itteration was exclusive, rather than inclusive. I was sort of passing the buck onto the person implementing the test to remove the directives that were causing them problems rather than have them add directives one at a time.

pgoforth committed:
chore: Add `frame-src` and `child-src` to conditional CSP directives

pgoforth committed:
chore: Rename `isSubsetOf` to `isArrayIncludingAny`

AtofStryker commented:
@pgoforth it might be a good idea then to make `child-src` and `frame-src` optional in the config. 

On another note, I got the CI working for contributor PRs again (sorry for the delay there) which took longer than anticipated. I got it merged in and some of the system tests are failing in a few browsers, so I will investigate that and fix the linting job. Heads up I will be committing directly to the branch

pgoforth commented:
@AtofStryker Looking at the test failures now. Fixing the typescript error and the snapshot that I forgot to update. I can also take a look at the system tests and see why those might be failing. Are those the ones you migrated?

Edit:
Oh...if you're already on it, I will let you fix it instead.

AtofStryker committed:
chore: fix CLI linting types

AtofStryker committed:
chore: fix server unit tests

AtofStryker commented:
@pgoforth no worries I got the server test and linting jobs fixed. I'm gonna focus on fixing the system tests since they might be a bit difficult to run and yes, those are the ones I migrated 😄 .

AtofStryker committed:
chore: fix system tests within firefox and webkit

AtofStryker commented:
@pgoforth I think I have the system tests fixed for webkit and firefox. Ill push up after the build is done running. Currently adding some more test coverage there for some of the optional directives

AtofStryker committed:
chore: add form-action test

AtofStryker commented:
@pgoforth OK I have the tests fixed and added a `form-action` test. I did experiment with `sandbox` and `navigate-to` and think we should always stripped these similar to `trusted-types` and `frame-ancestors` because
* sandbox pertains to all iframes in the document, regardless from where the request comes from (which seems really strange to me). In other words, this will almost definitely break cypress
* `navigate-to` isn't supported by a major browser yet and is in the CSP 3.0 draft. I am more on the side of always stripping it until its released and we know its intended side effects
would you be able to implement that? I know we have had a lot of churn on things but I think we are getting close 😄

pgoforth commented:
@AtofStryker Absolutely, that's a 2 line change.

AtofStryker committed:
chore: update system test snapshots

AtofStryker committed:
chore: skip tests in webkit due to form-action flakiness

pgoforth committed:
chore: Move 'sandbox' and 'navigate-to' into `unsupportedCSPDirectives`
- Add additional system tests
- Update snapshots and unit test

pgoforth commented:
@AtofStryker Moved the two directives into `unsupportedCSPDirectives` and added 2 additional system tests for them, leveraging the same form submission test you were using in another system test. Feel free to modify.

pgoforth commented:
@AtofStryker Looks like the snapshots for the new system tests I wrote need updating, and I had to update a unit test. I do not know how to run the system tests and update the snapshots. I'm rebasing off `develop`, resolving the conflict, and pushing up the unit test fix. Hopefully you can get the snapshots updated and passing.

AtofStryker commented:
@pgoforth taking a look and should have it updated soon

AtofStryker committed:
chore: update system test snapshots

AtofStryker committed:
chore: fix system tests

AtofStryker committed:
chore: do not run csp tests within firefox or webkit due to flake issues in CI

AtofStryker commented:
@emilyrohrbough I think this is ready for another look. I needed to tweak the system tests a bit but I think they provide ample coverage. We also added `sandbox` and `navigate-to` to the always strip directive since `sandbox` seems to apply to all iframes, which breaks Cypress, and `navigate-to` is not implemented in any browser yet and could very like cause issues with app navigation since in some cases we interact with the document directly.

AtofStryker committed:
chore: attempt to increase intercept delay to avoid race condition

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
Merge branch 'develop' into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
chore: update new snapshots with video defaults work

AtofStryker committed:
chore: update changelog

AtofStryker commented:
@pgoforth we are working on getting a few additional reviewers for this PR on Monday. TLDR there is a test that is [failing](https://app.circleci.com/pipelines/github/cypress-io/cypress/53469/workflows/81456162-c4cc-4fab-b3e7-d9a1dea83253/jobs/2205056/tests#failed-test-0) in CI on your branch, but we believe it is due to flake since the same commits on a [different branch](https://app.circleci.com/pipelines/github/cypress-io/cypress/53442/workflows/5772f6a7-98a9-42ac-bbb2-1806a0cf9065) owned by cypress ICs seems to work fine. This should be going into `develop` next week barring anything that might be problematic that is caught in review.

pgoforth commented:
@AtofStryker That all sounds great. I'm OOO next week, but will have my machine with me while I'm traveling. Let me know if there's anything I can assist with and will do my best to help out.

mschile committed:
Merge branch 'develop' into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
Merge branch 'develop' into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
Merge branch 'develop' into issue-1030/pgoforth/load-site-witout-csp-header-stripping

AtofStryker committed:
Merge branch 'develop' into issue-1030/pgoforth/load-site-witout-csp-header-stripping
